### PR TITLE
Fixes 172 and 152

### LINKED ISSUE #152: Limit users that have access to labman

### LINKED ISSUE #172: Blank/control values need more flexible typing
Currently, there are three preset control types:

 - blank:	gDNA extraction blanks. Represents an empty extraction well.
 - vibrio.positive.control:	Bacterial isolate control (Vibrio fischeri ES114). Represents an extraction well loaded with Vibrio.
 - zymo.mock:	Bacterial community control (Zymo Mock D6306). Represents an extraction well loaded with Zymo Mock community.

Hard coding this is probably a mistake, as there are different kinds of blanks (e.g. empty swab, extraction reagents, empty well) and likely many pore positive controls. 

Furthermore, anything can be coded as a sample and entered as positive information. Storing these controls is mostly 1) as a convenience, and 2) to allow API-assured access to certain types of control for downstream processing. 

Finally, in the plating interface, the behavior of these blanks should be more expicit -- i.e. currently, empty wells are designated blanks upon storing, but not visually prior. A 'fill blanks' button or something might do.

### PR DESCRIPTION
**Pull request by josenavas**
Fixes #172 
Fixes #152 

I want to get the tests kicked off - I'm working on adding an interface to let a user give access permissions as well as an interface to create controls.

### CODE REVIEW

**Comment by josenavas** [comment]
This is ready for review/merge

File: labman/gui/templates/access.html [hunk 1]
```diff
@@ -0,0 +1,91 @@
+{% extends sitebase.html %}
+
+{% block head %}
+<script type='text/javascript'>
+  function removeUser(email, name) {
+    $.post('/auth/access/', {'email': email, 'operation': 'revoke'}, function (data){
+      // Remove contact from additional contact list.
+      $('[id="user-' + email + '"]').remove();
```
Comment (ElDeveloper) [inline labman/gui/templates/access.html]: Minor but, the idiomatic way to search by id is by using `#` i.e. `$('#user-' + email).remove()`. Alternatively is there any reason why you are doing the search like this?

File: labman/gui/templates/plate.html [hunk 2]
```diff
@@ -131,18 +131,12 @@ <h4>Plate map</h4>
...
-      </tr>
-      <tr>
-        <td>vibrio.positive.control: </td>
-        <td> Bacterial isolate control (Vibrio fischeri ES114). Represents an extraction well loaded with Vibrio.</td>
-      </tr>
-      <tr>
-        <td>zymo.mock:</td>
-        <td>Bacterial community control (Zymo Mock D6306). Represents an extraction well loaded with Zymo Mock community.</td>
-      </tr>
+      {% for cd in controls_description %}
```
Comment (ElDeveloper) [inline labman/gui/templates/plate.html]: I would add a note at the foot of this table along the lines of: "To add new blank/sample types go to Admin>Manage control sample types" (perhaps even link to it).

File: labman/gui/templates/controls.html [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+    if (extId.replace(/\s/g,'').length === 0) {
+      extId = '';
+    }
+
+    if (description.replace(/\s/g,'').length === 0) {
+      description = '';
+    }
+
+    // Replace all non-alpha numeric characters by '.'
+    extId = extId.replace(/[^a-z0-9]/gmi, ".");
```
Comment (wasade) [inline labman/gui/templates/controls.html]: Is this case-insensitive?

File: labman/gui/handlers/auth.py [hunk 4]
```diff
@@ -47,3 +51,24 @@ class LogoutHandler(BaseHandler):
     def get(self):
         self.clear_cookie("user")
         self.redirect("/")
+
+
+class AccessHandler(BaseHandler):
+    @authenticated
+    def get(self):
+        self.render('access.html', users=User.list_users(),
+                    access_users=User.list_users(access_only=True))
```
Comment (wasade) [inline labman/gui/handlers/auth.py]: Does this only show users that have access? Is it possible to grant access to a user that isn't already on the list?

File: labman/gui/handlers/auth.py [hunk 5]
```diff
@@ -29,6 +31,8 @@ def post(self):
             error_msg = "Unknown user name"
         except LabmanLoginError:
             error_msg = "Incorrect password"
+        except LabmanLoginDisabledError:
+            error_msg = "User not allowed on this portal"
```
Comment (wasade) [inline labman/gui/handlers/auth.py]: There are multiple portals into labman, or just that the user doesn't have access to labman?

File: labman/db/tests/test_process.py [hunk 6]
```diff
@@ -1019,15 +1024,15 @@ def test_generate_echo_picklist(self):
         self.assertEqual(obs_lines[1],
                          '1,384LDV_AQ_B2_HT,A1,,1.00,NormalizedDNA,A1')
         self.assertEqual(obs_lines[-1],
-                         '1,384LDV_AQ_B2_HT,P24,,1.00,NormalizedDNA,A1')
+                         '1,384LDV_AQ_B2_HT,P24,,0.00,NormalizedDNA,A1')
```
Comment (wasade) [inline labman/db/tests/test_process.py]: Any idea why this changed?

File: labman/db/tests/test_composition.py [hunk 7]
```diff
@@ -320,21 +340,45 @@ def test_attributes(self):
 
     def test_get_next_combos(self):
         tester = ShotgunPrimerSet(1)
-        self.assertEqual(tester.current_combo_index, 384)
+        self.assertEqual(tester.current_combo_index, 380)
```
Comment (wasade) [inline labman/db/tests/test_composition.py]: It's not clear to me why this value changed, same with the one on line 311, would it be possible to provide some context?

File: labman/db/support_files/db_patch_manual.sql [hunk 8]
```diff
@@ -2,6 +2,14 @@
 CREATE INDEX composition_notes_idx ON qiita.composition USING gin(to_tsvector('english', notes));
 CREATE INDEX plate_notes_idx ON qiita.composition USING gin(to_tsvector('english', notes));
 
+-- Add a column on the qiita_user table to enable access for labmanager
+ALTER TABLE qiita.qiita_user ADD COLUMN labmanager_access bool DEFAULT 'FALSE';
```
Comment (wasade) [inline labman/db/support_files/db_patch_manual.sql]: Why not just create a table that has an FK relationship to `qiita.qiita_user`? There are easy possibilities for unintended consequences here, for instance, any method which previously performed `select * from qiita.qiita_user` is now getting a very different structure back.

cc @antgonza

**Comment by tanaes** [comment]
Not sure if this is related to the PR but when I pulled it down and tested out the interface, I noticed there were some problems resetting highlighting: 

feb-06-2018 09-32-44

File: labman/gui/templates/plate.html (hunk 2 above)
Comment (josenavas) [reply labman/gui/templates/plate.html]: Done

File: labman/db/composition.py [hunk 9]
```diff
@@ -399,6 +399,24 @@ class SampleComposition(Composition):
     _id_column = 'sample_composition_id'
     _composition_type = 'sample'
 
+    @staticmethod
+    def create_control_sample(external_id, description):
```
[the remark this answers is not in the document]
Comment (josenavas) [reply labman/db/composition.py]: control sample types - modified the function name for clarity

File: labman/db/composition.py [hunk 10]
```diff
@@ -417,16 +435,32 @@ def get_control_samples(term=None):
...
-                     WHERE description != 'experimental sample'
+                     WHERE external_id != 'experimental sample'
                      {}
-                     ORDER BY description""".format(sql_term)
+                     ORDER BY external_id""".format(sql_term)
             TRN.add(sql, sql_args)
             return TRN.execute_fetchflatten()
 
+    @staticmethod
+    def get_control_samples_description():
```
[the remark this answers is not in the document]
Comment (josenavas) [reply labman/db/composition.py]: control sample types, changed for clarity

**Comment by josenavas** [comment]
@tanaes I have not modified anything that would affect that behavior that you're seeing. The issue is unrelated with the changes performed in here.

**Comment by tanaes** [comment]
Ok cool will raise as issue

**Comment by tanaes** [comment]
@wasade any additional concerns? If not I can merge.

**Comment by josenavas** [comment]
Thanks @tanaes - @wasade @ElDeveloper can either of you review/merge if there are no more comments?

Thanks!

**Pull request merged**